### PR TITLE
Fix validation error in Add operation

### PR DESCRIPTION
**Pull request by reaganlo**
Fix issue #144

Signed-off-by: Reagan Lopez <reagan.lopez@intel.com>

### CODE REVIEW

**Comment by reaganlo** [comment]
@anikulk Please review and merge if ok.

**Comment by Nagamani71** [comment]
This change gives VTS memory test failure

**Comment by reaganlo** [comment]
@Nagamani71 Yes you are correct. Although it passes the ml_cmdline test, this change causes those VTS tests to fail.
```
localhost ~ # cros_nnapi_vts_1_3 --gtest_filter=*TestMemoryDomain/MemoryDomainAllocateTest.ConflictRankBetweenRoles*
Note: Google Test filter = *TestMemoryDomain/MemoryDomainAllocateTest.ConflictRankBetweenRoles*
[==========] Running 4 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 4 tests from TestMemoryDomain/MemoryDomainAllocateTest
[ RUN      ] TestMemoryDomain/MemoryDomainAllocateTest.ConflictRankBetweenRoles/cros_default_TENSOR_FLOAT32
Segmentation fault (core dumped)

localhost ~ # cros_nnapi_vts_1_3 --gtest_filter=*TestMemoryDomain/MemoryDomainAllocateTest.ConflictRankWithScalarRole*
Note: Google Test filter = *TestMemoryDomain/MemoryDomainAllocateTest.ConflictRankWithScalarRole*
[==========] Running 4 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 4 tests from TestMemoryDomain/MemoryDomainAllocateTest
[ RUN      ] TestMemoryDomain/MemoryDomainAllocateTest.ConflictRankWithScalarRole/cros_default_TENSOR_FLOAT32
Segmentation fault (core dumped)
```

**Comment by anikulk** [comment]
Hi Nagamani,

What is you original code checking for? Following are the different inputs for ADD

0: A tensor.
1: A tensor of the same [OperandCode](https://developer.android.com/ndk/reference/group/neural-networks#group___neural_networks_1gaf06d1affd33f3bc698d0c04eceb23298), and compatible dimensions as input0. For a [ANEURALNETWORKS_TENSOR_QUANT8_ASYMM](https://developer.android.com/ndk/reference/group/neural-networks#group___neural_networks_1ggaf06d1affd33f3bc698d0c04eceb23298a07984961d5c7c12f0f8c811bedd85dc3) and [ANEURALNETWORKS_TENSOR_QUANT8_ASYMM_SIGNED](https://developer.android.com/ndk/reference/group/neural-networks#group___neural_networks_1ggaf06d1affd33f3bc698d0c04eceb23298a0f7c57dd01611105bbdd17922dacbba9) tensor, the scales and zeroPoint can be different from input0 scale and zeroPoint.
2: An [ANEURALNETWORKS_INT32](https://developer.android.com/ndk/reference/group/neural-networks#group___neural_networks_1ggaf06d1affd33f3bc698d0c04eceb23298acb165f692e9d189c820ceba84b3c21dc) scalar, and has to be one of the [FuseCode](https://developer.android.com/ndk/reference/group/neural-networks#group___neural_networks_1ga4ec620045a39909a6bf00b3ee0c4d414) values. Specifies the activation to invoke on the result. For a [ANEURALNETWORKS_TENSOR_INT32](https://developer.android.com/ndk/reference/group/neural-networks#group___neural_networks_1ggaf06d1affd33f3bc698d0c04eceb23298ac34965d8e76ac5acfddf5acd9e40f896) tensor, the [FuseCode](https://developer.android.com/ndk/reference/group/neural-networks#group___neural_networks_1ga4ec620045a39909a6bf00b3ee0c4d414) must be "NONE".


0 and 1 are inputs so we cannot force them to be CONST. 2 is the activation Code.

**Comment by Nagamani71** [comment]
Hi @anikulk, initially I have added those checkes to eliminate unsupported tensor types. After adding support for all tensors i have removed that check in nn-hal 1.2
However, with 1.3 new memory domain VTS test cases are added, after debugging the issue added back this check in Add operation (https://github.com/intel/nn-hal/pull/142/commits/e49043b8d25c43242191c1f1b111c5ef9605fc88)

**Comment by reaganlo** [comment]
@Nagamani71 Based on Anisha's suggestion, do you see any issues if we change the check to get the activation code from idx 2
`const auto& activationIndex = sModelInfo->getOperationInput(mNnapiOperationIndex, 2);`

It passes the ml_cmdline test as well as the 2 memory tests that you mentioned.

Please let us know.

**Comment by anikulk** [comment]
@Nagamani71 

Its not clear what issue the test case exposes. In my understanding "somehow" this validation check avoided the memorydomaintest fail, but is not the actual fix .   Can you clarify? Parallelly I am looking into to test.

Anisha

**Comment by Nagamani71** [comment]
I created PR with wrong validation index, just noticed. It supposed to be index 2. 
Issue is: Activation function supposed to be a scalar and we are getting scalar data from "ParseOperationInput" function which internally calls "GetConstFromBuffer" (function to fetch value from the buffer). However, VTS MemoryDomainAllocateTest test cases has Activation function as "SUBGRAPH_INPUT" (buffer is "null") hence, we are getting "Cause: null pointer dereference" error. To eliminate such cases I have added that validation check.

**Comment by anikulk** [comment]
@Nagamani71 Being scalar(OperationType INT32) and being a CONST lifetime are different restrictions.

**Comment by Nagamani71** [comment]
@anikulk I agree, but here problem is with scalar (FusedActivation) having lifetime "SUBGRAPH_INPUT" (which is not supported by our framework).

**Comment by anikulk** [comment]
Hi @reaganlo ,

You can use the change suggested, i.e

const auto& activationIndex = sModelInfo->getOperationInput(mNnapiOperationIndex, 2);

Can you also make sure that the error message  for the fail captures that this is a restriction from our HAL. i.e due to OpenVINO API restrictions Scalar input values have to be CONST lifetime. 
Essentially we will skip the case (both in product and test cases) when activation is non CONST.

File: ngraph_creator/operations/src/Add.cpp [hunk 1]
```diff
@@ -14,12 +14,6 @@ Add::Add(int operationIndex) : OperationsBase(operationIndex) {
 bool Add::validate() {
     ALOGV("%s PASSED", __func__);
 
-    const auto& activationIndex = sModelInfo->getOperationInput(mNnapiOperationIndex, 1);
```
Comment (anikulk) [inline ngraph_creator/operations/src/Add.cpp:17]: Change index to 2

File: ngraph_creator/operations/src/Add.cpp [hunk 2]
```diff
@@ -14,12 +14,6 @@ Add::Add(int operationIndex) : OperationsBase(operationIndex) {
...
-    if (!sModelInfo->isOperandLifeTimeConst(activationIndex)) {
-        ALOGE("%s Only Constant supported for specifying Activation", __func__);
```
Comment (anikulk) [inline ngraph_creator/operations/src/Add.cpp:19]: More descriptive error msg to capture why this validation check exists.

**Comment by reaganlo** [comment]
```
localhost ~ # ml_cmdline --nnapi
Adding 1 and 4 with NNAPI
INFO: Created TensorFlow Lite delegate for NNAPI.
2022-03-07T14:51:10.083384Z INFO ml_cmdline: [logger.cpp(31)] Manager:Manager.cpp:1285: DeviceManager::DeviceManager
2022-03-07T14:51:10.083435Z INFO ml_cmdline: [logger.cpp(31)] Manager:Manager.cpp:1262: findAvailableDevices
2022-03-07T14:51:10.083444Z INFO ml_cmdline: [logger.cpp(31)] ml_cmdline:driver_loader.cpp:43: Loading service default
2022-03-07T14:51:10.083479Z INFO ml_cmdline: [logger.cpp(31)] Driver::0: CPU driver getCapabilities()
2022-03-07T14:51:10.083489Z INFO ml_cmdline: [logger.cpp(31)] Driver::0: CPU MKLDNN driver Capabilities .execTime = 0.9f, .powerUsage = 0.9f
2022-03-07T14:51:10.083513Z INFO ml_cmdline: [logger.cpp(31)] ml_cmdline:driver_loader.cpp:51: Creating SharedDevice for default
2022-03-07T14:51:10.083519Z INFO ml_cmdline: [logger.cpp(31)] Manager:Manager.cpp:1267: Found interface default
2022-03-07T14:51:10.083807Z INFO ml_cmdline: [logger.cpp(31)] ModelManager::0: Operand (0) {.type = TENSOR_FLOAT32, .dimensions = [1]{1}, .numberOfConsumers = 1, .scale = 0.000000, .zeroPoint = 0, .lifetime = SUBGRAPH_INPUT, .location = {.poolIndex = 0, .offset = 0, .length = 0}, .extraParams = {.none = {}}}
2022-03-07T14:51:10.083821Z INFO ml_cmdline: [logger.cpp(31)] ModelManager::0: Operand (1) {.type = TENSOR_FLOAT32, .dimensions = [1]{1}, .numberOfConsumers = 1, .scale = 0.000000, .zeroPoint = 0, .lifetime = SUBGRAPH_INPUT, .location = {.poolIndex = 0, .offset = 0, .length = 0}, .extraParams = {.none = {}}}
2022-03-07T14:51:10.083828Z INFO ml_cmdline: [logger.cpp(31)] ModelManager::0: Operand (2) {.type = INT32, .dimensions = [0]{}, .numberOfConsumers = 1, .scale = 0.000000, .zeroPoint = 0, .lifetime = CONSTANT_COPY, .location = {.poolIndex = 0, .offset = 0, .length = 4}, .extraParams = {.none = {}}}
2022-03-07T14:51:10.083834Z INFO ml_cmdline: [logger.cpp(31)] ModelManager::0: Operand (3) {.type = TENSOR_FLOAT32, .dimensions = [1]{1}, .numberOfConsumers = 0, .scale = 0.000000, .zeroPoint = 0, .lifetime = SUBGRAPH_OUTPUT, .location = {.poolIndex = 0, .offset = 0, .length = 0}, .extraParams = {.none = {}}}
2022-03-07T14:51:10.083840Z INFO ml_cmdline: [logger.cpp(31)] CpuPreparedModel::0: Generating IR Graph
2022-03-07T14:51:10.089223Z INFO ml_cmdline: [logger.cpp(31)] ModelManager::0: getBlobFromMemoryPoolIn Operand length:4 pointer:0x78e467887000 offset:0 pool index: 0
2022-03-07T14:51:10.089241Z INFO ml_cmdline: [logger.cpp(31)] ModelManager::0: getBlobFromMemoryPoolIn Operand length:4 pointer:0x78e467887040 offset:64 pool index: 0
2022-03-07T14:51:10.089245Z INFO ml_cmdline: [logger.cpp(31)] IENetwork::0: Infer Network

2022-03-07T14:51:10.089395Z INFO ml_cmdline: [logger.cpp(31)] IENetwork::0: infer request completed
2022-03-07T14:51:10.089399Z INFO ml_cmdline: [logger.cpp(31)] BasePreparedModel::0: OutputIndex: 3
2022-03-07T14:51:10.089405Z INFO ml_cmdline: [logger.cpp(31)] ModelManager::0: getBlobFromMemoryPoolOut Operand length:4 pointer:0x78e467886000
Status: OK
Sum: 5
localhost ~ # cros_nnapi_vts_1_3 --gtest_filter=*TestMemoryDomain/MemoryDomainAllocateTest.ConflictRankBetweenRoles*
Note: Google Test filter = *TestMemoryDomain/MemoryDomainAllocateTest.ConflictRankBetweenRoles*
[==========] Running 4 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 4 tests from TestMemoryDomain/MemoryDomainAllocateTest
[ RUN      ] TestMemoryDomain/MemoryDomainAllocateTest.ConflictRankBetweenRoles/cros_default_TENSOR_FLOAT32
[       OK ] TestMemoryDomain/MemoryDomainAllocateTest.ConflictRankBetweenRoles/cros_default_TENSOR_FLOAT32 (0 ms)
[ RUN      ] TestMemoryDomain/MemoryDomainAllocateTest.ConflictRankBetweenRoles/cros_default_TENSOR_FLOAT16
[       OK ] TestMemoryDomain/MemoryDomainAllocateTest.ConflictRankBetweenRoles/cros_default_TENSOR_FLOAT16 (0 ms)
[ RUN      ] TestMemoryDomain/MemoryDomainAllocateTest.ConflictRankBetweenRoles/cros_default_TENSOR_QUANT8_ASYMM
[       OK ] TestMemoryDomain/MemoryDomainAllocateTest.ConflictRankBetweenRoles/cros_default_TENSOR_QUANT8_ASYMM (0 ms)
[ RUN      ] TestMemoryDomain/MemoryDomainAllocateTest.ConflictRankBetweenRoles/cros_default_TENSOR_QUANT8_ASYMM_SIGNED
[       OK ] TestMemoryDomain/MemoryDomainAllocateTest.ConflictRankBetweenRoles/cros_default_TENSOR_QUANT8_ASYMM_SIGNED (0 ms)
[----------] 4 tests from TestMemoryDomain/MemoryDomainAllocateTest (0 ms total)

[----------] Global test environment tear-down
[==========] 4 tests from 1 test suite ran. (0 ms total)
[  PASSED  ] 4 tests.
localhost ~ # cros_nnapi_vts_1_3 --gtest_filter=*TestMemoryDomain/MemoryDomainAllocateTest.ConflictRankWithScalarRole*
Note: Google Test filter = *TestMemoryDomain/MemoryDomainAllocateTest.ConflictRankWithScalarRole*
[==========] Running 4 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 4 tests from TestMemoryDomain/MemoryDomainAllocateTest
[ RUN      ] TestMemoryDomain/MemoryDomainAllocateTest.ConflictRankWithScalarRole/cros_default_TENSOR_FLOAT32
[       OK ] TestMemoryDomain/MemoryDomainAllocateTest.ConflictRankWithScalarRole/cros_default_TENSOR_FLOAT32 (0 ms)
[ RUN      ] TestMemoryDomain/MemoryDomainAllocateTest.ConflictRankWithScalarRole/cros_default_TENSOR_FLOAT16
[       OK ] TestMemoryDomain/MemoryDomainAllocateTest.ConflictRankWithScalarRole/cros_default_TENSOR_FLOAT16 (0 ms)
[ RUN      ] TestMemoryDomain/MemoryDomainAllocateTest.ConflictRankWithScalarRole/cros_default_TENSOR_QUANT8_ASYMM
[       OK ] TestMemoryDomain/MemoryDomainAllocateTest.ConflictRankWithScalarRole/cros_default_TENSOR_QUANT8_ASYMM (0 ms)
[ RUN      ] TestMemoryDomain/MemoryDomainAllocateTest.ConflictRankWithScalarRole/cros_default_TENSOR_QUANT8_ASYMM_SIGNED
[       OK ] TestMemoryDomain/MemoryDomainAllocateTest.ConflictRankWithScalarRole/cros_default_TENSOR_QUANT8_ASYMM_SIGNED (0 ms)
[----------] 4 tests from TestMemoryDomain/MemoryDomainAllocateTest (0 ms total)

[----------] Global test environment tear-down
[==========] 4 tests from 1 test suite ran. (1 ms total)
[  PASSED  ] 4 tests.
```

File: ngraph_creator/operations/src/Add.cpp (hunk 1 above)
Comment (reaganlo) [reply ngraph_creator/operations/src/Add.cpp:17]: @anikulk Done

File: ngraph_creator/operations/src/Add.cpp (hunk 2 above)
Comment (reaganlo) [reply ngraph_creator/operations/src/Add.cpp:19]: @anikulk Done

**Pull request merged**